apps/utils/langchain.py

```diff
@@ -155,6 +174,21 @@ def _call(self, messages: list[BaseMessage], *args, **kwargs) -> str | BaseMessa
 
         return f"{system_message} {user_message}" if self.include_system_message else user_message
 
+    def with_structured_output(self, schema) -> RunnableSerializable:
+        """Creates a runnable that returns objects with a 'route' attribute for router schemas."""
+
+        def _structured_output_handler(input_value, *args, **kwargs):
+            if isinstance(input_value, ChatPromptValue):
+                messages = input_value.messages
+            else:
+                messages = [input_value]
+            self._call(messages, *args, **kwargs)
+            user_message = messages[-1].content.lower() if messages else ""
+            if hasattr(schema, "__annotations__") and "route" in schema.__annotations__:
+                return type("RouterOutput", (), {"route": user_message})
+
+        return RunnableLambda(_structured_output_handler)
```

Comment:
do you need to re-define this on this class? Can't you put this implementation on the base `FakeLlm` class?

Reply:
[de59d0a](https://github.com/dimagi/open-chat-studio/pull/1403/commits/de59d0a39e4a3a0c6633f0415f26845a8b0ff089) right thanks